src/main/java/org/apache/sysml/utils/NativeHelper.java

```diff
@@ -0,0 +1,256 @@
+			os = "apple";
+		} else if (SystemUtils.IS_OS_LINUX) {
+			prefix = "lib";
+			suffix2 = "so";
+			os = "linux";
+		} else if (SystemUtils.IS_OS_WINDOWS) {
+			prefix = "";
+			suffix2 = "dll";
+			os = "windows";
+		} else {
+			LOG.info("Unsupported OS:" + SystemUtils.OS_NAME);
+			throw new IOException("Unsupported OS");
+		}
+		
+		String arch = archMap.get(SystemUtils.OS_ARCH);
+		if(arch == null) {
+			LOG.info("Unsupported architecture:" + SystemUtils.OS_ARCH);
+			throw new IOException("Unsupported architecture:" + SystemUtils.OS_ARCH);
+		}
+		loadLibraryHelper(prefix + libName + suffix1 + "-" + os + "-" + arch + "." + suffix2);
+	}
+
+	private static void loadLibraryHelper(String path) throws IOException {
+		InputStream in = null; OutputStream out = null;
+		try {
+			in = NativeHelper.class.getResourceAsStream("/lib/"+path);
+			if(in != null) {
+				File temp = File.createTempFile(path, "");
+				temp.deleteOnExit();
+				out = FileUtils.openOutputStream(temp);
+		        IOUtils.copy(in, out);
+		        in.close(); in = null;
+		        out.close(); out = null;
+				System.load(temp.getAbsolutePath());
+			}
+			else
+				throw new IOException("No lib available in the jar:" + path);
+			
+		} catch(IOException e) {
+			LOG.info("Unable to load library " + path + " from resource:" + e.getMessage());
+			throw e;
+		} finally {
+			if(out != null)
+				out.close();
+			if(in != null)
+				in.close();
+		}
+		
+	}
+	
+	// TODO: Add pmm, wsloss, mmchain, etc.
+	public static native boolean matrixMultDenseDense(double [] m1, double [] m2, double [] ret, int m1rlen, int m1clen, int m2clen, int numThreads);
+	public static native boolean tsmm(double [] m1, double [] ret, int m1rlen, int m1clen, boolean isLeftTranspose, int numThreads);
+	
+	// LibMatrixDNN operations:
+	// N = number of images, C = number of channels, H = image height, W = image width
+	// K = number of filters, R = filter height, S = filter width
+	public static native boolean conv2dDense(double [] input, double [] filter, double [] ret, int N, int C, int H, int W, 
+			int K, int R, int S, int stride_h, int stride_w, int pad_h, int pad_w, int P, int Q, int numThreads);
+	public static native boolean conv2dBiasAddDense(double [] input, double [] bias, double [] filter, double [] ret, int N, int C, int H, int W, 
+			int K, int R, int S, int stride_h, int stride_w, int pad_h, int pad_w, int P, int Q, int numThreads);
+	public static native boolean conv2dBackwardDataDense(double [] filter, double [] dout, double [] ret, int N, int C, int H, int W, 
+			int K, int R, int S, int stride_h, int stride_w, int pad_h, int pad_w, int P, int Q, int numThreads);
+	public static native boolean conv2dBackwardFilterDense(double [] input, double [] dout, double [] ret, int N, int C, int H, int W, 
+			int K, int R, int S, int stride_h, int stride_w, int pad_h, int pad_w, int P, int Q, int numThreads);
+	
+	private static native void setMaxNumThreads(int numThreads);
```

Comment:
Does the current lazy initialization approach for the maximum number of threads handle the case of changing constraints as used during parfor (e.g., we start with 24 outside parfor and 24/k inside parfor where k is the parfor degree of parallelism)